pandas/core/resample.py

```diff
@@ -1348,18 +1352,32 @@ def __init__(
         self.convention = convention or "E"
         self.convention = self.convention.lower()
 
-        if isinstance(loffset, str):
-            loffset = to_offset(loffset)
-        self.loffset = loffset
-
         self.how = how
         self.fill_method = fill_method
         self.limit = limit
-        self.base = base
+
+        self.origin = Timestamp(origin) if origin is not None else None
```

Comment:
Related to comment on annotations above, do we want to wrap things like this? Seems a little out of place to do here; I think should explicitly require Timestamp / Timedelta instances or pass through to other functions that will box accordingly

Reply:
I think what we need is to add something like this in `pandas/_typing.py`

(based on the documentation 
```python
# based on the doc and on pandas._libs.tslibs.timestamps.pyx::convert_to_tsobject
TimestampCompatibleType = Union[
    "Timestamp", datetime, np.datetime64, int, np.int64, float, str
]

# based on the doc and on pandas._libs.tslibs.timedeltas.pyx::convert_to_timedelta64
TimedeltaCompatibleType = Union[
    "Timedelta", timedelta, np.timedelta64, int, np.int64, float, str
]
```

And we will have in the constructor something like this
```python
origin: Optional[TimestampCompatibleType] = None,
offset: Optional[TimedeltaCompatibleType] = None,
```

The interest is really to handle origin/offset as string or any compatible type of Timestamp/Timedelta.

That kind of things would be supported out of the box (and I think this is what we want):
```python
ts.resample("17min", origin=dt.datetime(year=1993, month=1, day=5)).sum()
ts.resample("17min", origin="1993-01-05").sum()
ts.resample("17min", offset="16h30").sum()
```

I have started the implementation of this, I'll finish that soon. :wink:

Reply:
I have annotated the arguments `offset` and `origin` (in the last commit).

> Seems a little out of place to do here; I think should explicitly require Timestamp / Timedelta instances
To be honest I think doing the following is fine:

```python
self.origin = Timestamp(origin) if origin is not None else None
```
It is done elsewhere in the code. If I have to take only one example, here is a piece of code that handle the arguments of `pd.date_range`:

https://github.com/pandas-dev/pandas/blob/a6a1ab20e1fd3b9c54308f852a4427846b0695da/pandas/core/arrays/datetimes.py#L364

and the doc is quite similar:
```
start : str or datetime-like, optional
```

> or pass through to other functions that will box accordingly

I have checked the code of Timestamp, and is seems the constructor __new__  is acting like a "boxing" function since it does not return a copy of `ts_input` if `ts_input` is already a Timestamp:
https://github.com/pandas-dev/pandas/blob/a6a1ab20e1fd3b9c54308f852a4427846b0695da/pandas/_libs/tslibs/timestamps.pyx#L427

Same thing for Timedelta:
https://github.com/pandas-dev/pandas/blob/a6a1ab20e1fd3b9c54308f852a4427846b0695da/pandas/_libs/tslibs/timedeltas.pyx#L1232

It seems the related issues/PR are recent though :slightly_smiling_face: https://github.com/pandas-dev/pandas/issues/30543